#### Welcome to second question of assignment.

# Question:2 : Toronto postal codes with coordinates 

#### Here we continue from the dataframes for Toronto postal codes. Then we create next dataframe which has coordinates and then will merge two dataframes which shows postal code, borough, neighbourhood, latitude and longitude in single dataframe.

### Converting html table to panda dataframe

#### The html table is first extracted from the Wikipedia page. The html table is stored in a variable html2. Using BeatifulSoup I am trying to convert HTML table into panda dataframe.


In [3]:
# Import required libraries
import pandas as pd
from bs4 import BeautifulSoup
print("Done")

Done


In [4]:

html2 = """
<table class="wikitable">
<tr>
<th>Postal code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park / Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor / Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park / Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern / Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3B
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4B
</td>
<td>East York
</td>
<td>Parkview Hill / Woodbine Gardens
</td></tr>
<tr>
<td>M5B
</td>
<td>Downtown Toronto
</td>
<td>Garden District, Ryerson
</td></tr>
<tr>
<td>M6B
</td>
<td>North York
</td>
<td>Glencairn
</td></tr>
<tr>
<td>M7B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9B
</td>
<td>Etobicoke
</td>
<td>West Deane Park / Princess Gardens / Martin Grove / Islington / Cloverdale
</td></tr>
<tr>
<td>M1C
</td>
<td>Scarborough
</td>
<td>Rouge Hill / Port Union / Highland Creek
</td></tr>
<tr>
<td>M2C
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3C
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4C
</td>
<td>East York
</td>
<td>Woodbine Heights
</td></tr>
<tr>
<td>M5C
</td>
<td>Downtown Toronto
</td>
<td>St. James Town
</td></tr>
<tr>
<td>M6C
</td>
<td>York
</td>
<td>Humewood-Cedarvale
</td></tr>
<tr>
<td>M7C
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8C
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9C
</td>
<td>Etobicoke
</td>
<td>Eringate / Bloordale Gardens / Old Burnhamthorpe / Markland Wood
</td></tr>
<tr>
<td>M1E
</td>
<td>Scarborough
</td>
<td>Guildwood / Morningside / West Hill
</td></tr>
<tr>
<td>M2E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4E
</td>
<td>East Toronto
</td>
<td>The Beaches
</td></tr>
<tr>
<td>M5E
</td>
<td>Downtown Toronto
</td>
<td>Berczy Park
</td></tr>
<tr>
<td>M6E
</td>
<td>York
</td>
<td>Caledonia-Fairbanks
</td></tr>
<tr>
<td>M7E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1G
</td>
<td>Scarborough
</td>
<td>Woburn
</td></tr>
<tr>
<td>M2G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4G
</td>
<td>East York
</td>
<td>Leaside
</td></tr>
<tr>
<td>M5G
</td>
<td>Downtown Toronto
</td>
<td>Central Bay Street
</td></tr>
<tr>
<td>M6G
</td>
<td>Downtown Toronto
</td>
<td>Christie
</td></tr>
<tr>
<td>M7G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1H
</td>
<td>Scarborough
</td>
<td>Cedarbrae
</td></tr>
<tr>
<td>M2H
</td>
<td>North York
</td>
<td>Hillcrest Village
</td></tr>
<tr>
<td>M3H
</td>
<td>North York
</td>
<td>Bathurst Manor / Wilson Heights / Downsview North
</td></tr>
<tr>
<td>M4H
</td>
<td>East York
</td>
<td>Thorncliffe Park
</td></tr>
<tr>
<td>M5H
</td>
<td>Downtown Toronto
</td>
<td>Richmond / Adelaide / King
</td></tr>
<tr>
<td>M6H
</td>
<td>West Toronto
</td>
<td>Dufferin / Dovercourt Village
</td></tr>
<tr>
<td>M7H
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8H
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9H
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1J
</td>
<td>Scarborough
</td>
<td>Scarborough Village
</td></tr>
<tr>
<td>M2J
</td>
<td>North York
</td>
<td>Fairview / Henry Farm / Oriole
</td></tr>
<tr>
<td>M3J
</td>
<td>North York
</td>
<td>Northwood Park / York University
</td></tr>
<tr>
<td>M4J
</td>
<td>East York
</td>
<td>East Toronto
</td></tr>
<tr>
<td>M5J
</td>
<td>Downtown Toronto
</td>
<td>Harbourfront East / Union Station / Toronto Islands
</td></tr>
<tr>
<td>M6J
</td>
<td>West Toronto
</td>
<td>Little Portugal / Trinity
</td></tr>
<tr>
<td>M7J
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8J
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9J
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1K
</td>
<td>Scarborough
</td>
<td>Kennedy Park / Ionview / East Birchmount Park
</td></tr>
<tr>
<td>M2K
</td>
<td>North York
</td>
<td>Bayview Village
</td></tr>
<tr>
<td>M3K
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4K
</td>
<td>East Toronto
</td>
<td>The Danforth West / Riverdale
</td></tr>
<tr>
<td>M5K
</td>
<td>Downtown Toronto
</td>
<td>Toronto Dominion Centre / Design Exchange
</td></tr>
<tr>
<td>M6K
</td>
<td>West Toronto
</td>
<td>Brockton / Parkdale Village / Exhibition Place
</td></tr>
<tr>
<td>M7K
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8K
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9K
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1L
</td>
<td>Scarborough
</td>
<td>Golden Mile / Clairlea / Oakridge
</td></tr>
<tr>
<td>M2L
</td>
<td>North York
</td>
<td>York Mills / Silver Hills
</td></tr>
<tr>
<td>M3L
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4L
</td>
<td>East Toronto
</td>
<td>India Bazaar / The Beaches West
</td></tr>
<tr>
<td>M5L
</td>
<td>Downtown Toronto
</td>
<td>Commerce Court / Victoria Hotel
</td></tr>
<tr>
<td>M6L
</td>
<td>North York
</td>
<td>North Park / Maple Leaf Park / Upwood Park
</td></tr>
<tr>
<td>M7L
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8L
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9L
</td>
<td>North York
</td>
<td>Humber Summit
</td></tr>
<tr>
<td>M1M
</td>
<td>Scarborough
</td>
<td>Cliffside / Cliffcrest / Scarborough Village West
</td></tr>
<tr>
<td>M2M
</td>
<td>North York
</td>
<td>Willowdale / Newtonbrook
</td></tr>
<tr>
<td>M3M
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4M
</td>
<td>East Toronto
</td>
<td>Studio District
</td></tr>
<tr>
<td>M5M
</td>
<td>North York
</td>
<td>Bedford Park / Lawrence Manor East
</td></tr>
<tr>
<td>M6M
</td>
<td>York
</td>
<td>Del Ray / Mount Dennis / Keelsdale and Silverthorn
</td></tr>
<tr>
<td>M7M
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8M
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9M
</td>
<td>North York
</td>
<td>Humberlea / Emery
</td></tr>
<tr>
<td>M1N
</td>
<td>Scarborough
</td>
<td>Birch Cliff / Cliffside West
</td></tr>
<tr>
<td>M2N
</td>
<td>North York
</td>
<td>Willowdale
</td></tr>
<tr>
<td>M3N
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4N
</td>
<td>Central Toronto
</td>
<td>Lawrence Park
</td></tr>
<tr>
<td>M5N
</td>
<td>Central Toronto
</td>
<td>Roselawn
</td></tr>
<tr>
<td>M6N
</td>
<td>York
</td>
<td>Runnymede / The Junction North
</td></tr>
<tr>
<td>M7N
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8N
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9N
</td>
<td>York
</td>
<td>Weston
</td></tr>
<tr>
<td>M1P
</td>
<td>Scarborough
</td>
<td>Dorset Park / Wexford Heights / Scarborough Town Centre
</td></tr>
<tr>
<td>M2P
</td>
<td>North York
</td>
<td>York Mills West
</td></tr>
<tr>
<td>M3P
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4P
</td>
<td>Central Toronto
</td>
<td>Davisville North
</td></tr>
<tr>
<td>M5P
</td>
<td>Central Toronto
</td>
<td>Forest Hill North &amp; West
</td></tr>
<tr>
<td>M6P
</td>
<td>West Toronto
</td>
<td>High Park / The Junction South
</td></tr>
<tr>
<td>M7P
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8P
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9P
</td>
<td>Etobicoke
</td>
<td>Westmount
</td></tr>
<tr>
<td>M1R
</td>
<td>Scarborough
</td>
<td>Wexford / Maryvale
</td></tr>
<tr>
<td>M2R
</td>
<td>North York
</td>
<td>Willowdale
</td></tr>
<tr>
<td>M3R
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4R
</td>
<td>Central Toronto
</td>
<td>North Toronto West
</td></tr>
<tr>
<td>M5R
</td>
<td>Central Toronto
</td>
<td>The Annex / North Midtown / Yorkville
</td></tr>
<tr>
<td>M6R
</td>
<td>West Toronto
</td>
<td>Parkdale / Roncesvalles
</td></tr>
<tr>
<td>M7R
</td>
<td>Mississauga
</td>
<td>Canada Post Gateway Processing Centre
</td></tr>
<tr>
<td>M8R
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9R
</td>
<td>Etobicoke
</td>
<td>Kingsview Village / St. Phillips / Martin Grove Gardens / Richview Gardens
</td></tr>
<tr>
<td>M1S
</td>
<td>Scarborough
</td>
<td>Agincourt
</td></tr>
<tr>
<td>M2S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4S
</td>
<td>Central Toronto
</td>
<td>Davisville
</td></tr>
<tr>
<td>M5S
</td>
<td>Downtown Toronto
</td>
<td>University of Toronto / Harbord
</td></tr>
<tr>
<td>M6S
</td>
<td>West Toronto
</td>
<td>Runnymede / Swansea
</td></tr>
<tr>
<td>M7S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1T
</td>
<td>Scarborough
</td>
<td>Clarks Corners / Tam O'Shanter / Sullivan
</td></tr>
<tr>
<td>M2T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4T
</td>
<td>Central Toronto
</td>
<td>Moore Park / Summerhill East
</td></tr>
<tr>
<td>M5T
</td>
<td>Downtown Toronto
</td>
<td>Kensington Market / Chinatown / Grange Park
</td></tr>
<tr>
<td>M6T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1V
</td>
<td>Scarborough
</td>
<td>Milliken / Agincourt North / Steeles East / L'Amoreaux East
</td></tr>
<tr>
<td>M2V
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3V
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4V
</td>
<td>Central Toronto
</td>
<td>Summerhill West / Rathnelly / South Hill / Forest Hill SE / Deer Park
</td></tr>
<tr>
<td>M5V
</td>
<td>Downtown Toronto
</td>
<td>CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst
<pre> Quay / South Niagara / Island airport
</pre>
</td></tr>
<tr>
<td>M6V
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7V
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8V
</td>
<td>Etobicoke
</td>
<td>New Toronto / Mimico South / Humber Bay Shores
</td></tr>
<tr>
<td>M9V
</td>
<td>Etobicoke
</td>
<td>South Steeles / Silverstone / Humbergate / Jamestown / Mount Olive / Beaumond Heights / Thistletown / Albion Gardens
</td></tr>
<tr>
<td>M1W
</td>
<td>Scarborough
</td>
<td>Steeles West / L'Amoreaux West
</td></tr>
<tr>
<td>M2W
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3W
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4W
</td>
<td>Downtown Toronto
</td>
<td>Rosedale
</td></tr>
<tr>
<td>M5W
</td>
<td>Downtown Toronto
</td>
<td>Stn A PO Boxes
</td></tr>
<tr>
<td>M6W
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7W
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8W
</td>
<td>Etobicoke
</td>
<td>Alderwood / Long Branch
</td></tr>
<tr>
<td>M9W
</td>
<td>Etobicoke
</td>
<td>Northwest
</td></tr>
<tr>
<td>M1X
</td>
<td>Scarborough
</td>
<td>Upper Rouge
</td></tr>
<tr>
<td>M2X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4X
</td>
<td>Downtown Toronto
</td>
<td>St. James Town / Cabbagetown
</td></tr>
<tr>
<td>M5X
</td>
<td>Downtown Toronto
</td>
<td>First Canadian Place / Underground city
</td></tr>
<tr>
<td>M6X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8X
</td>
<td>Etobicoke
</td>
<td>The Kingsway / Montgomery Road  / Old Mill North
</td></tr>
<tr>
<td>M9X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4Y
</td>
<td>Downtown Toronto
</td>
<td>Church and Wellesley
</td></tr>
<tr>
<td>M5Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M6Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7Y
</td>
<td>East Toronto
</td>
<td>Business reply mail Processing CentrE
</td></tr>
<tr>
<td>M8Y
</td>
<td>Etobicoke
</td>
<td>Old Mill South / King's Mill Park / Sunnylea / Humber Bay / Mimico NE / The Queensway East / Royal York South East / Kingsway Park South East
</td></tr>
<tr>
<td>M9Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M5Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M6Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8Z
</td>
<td>Etobicoke
</td>
<td>Mimico NW / The Queensway West / South of Bloor / Kingsway Park South West / Royal York South West
</td></tr>
<tr>
<td>M9Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
</table>"""





#### Use BeutifulSoup to convert html to dataframe.

In [5]:
soup = BeautifulSoup(html2, 'html.parser')
table = soup.find("table")

# The first tr contains the field names.
headings = [th.get_text().strip() for th in table.find("tr").find_all("th")]


datasets = []
for row in table.find_all("tr")[1:]:
    dataset = dict(zip(headings, (td.get_text() for td in row.find_all("td"))))
    datasets.append(dataset)


df=pd.DataFrame(datasets)
df.head(20)

Borough                                       Neighborhood  \
0       Not assigned\n                                                 \n   
1       Not assigned\n                                                 \n   
2         North York\n                                        Parkwoods\n   
3         North York\n                                 Victoria Village\n   
4   Downtown Toronto\n                       Regent Park / Harbourfront\n   
5         North York\n                Lawrence Manor / Lawrence Heights\n   
6   Downtown Toronto\n     Queen's Park / Ontario Provincial Government\n   
7       Not assigned\n                                                 \n   
8          Etobicoke\n                                 Islington Avenue\n   
9        Scarborough\n                                  Malvern / Rouge\n   
10      Not assigned\n                                                 \n   
11        North York\n                                        Don Mills\n   
12         East York\n                 Parkview Hill / Woodbine Gardens\n   
13  Downtown Toronto\n                         Garden District, Ryerson\n   
14        North York\n                                        Glencairn\n   
15      Not assigned\n                                                 \n   
16      Not assigned\n                                                 \n   
17         Etobicoke\n  West Deane Park / Princess Gardens / Martin Gr...   
18       Scarborough\n         Rouge Hill / Port Union / Highland Creek\n   
19      Not assigned\n                                                 \n   

   Postal code  
0        M1A\n  
1        M2A\n  
2        M3A\n  
3        M4A\n  
4        M5A\n  
5        M6A\n  
6        M7A\n  
7        M8A\n  
8        M9A\n  
9        M1B\n  
10       M2B\n  
11       M3B\n  
12       M4B\n  
13       M5B\n  
14       M6B\n  
15       M7B\n  
16       M8B\n  
17       M9B\n  
18       M1C\n  
19       M2C\n

### Remove last two unwanted characters

In [6]:
df['Borough'] = df['Borough'].map(lambda x: str(x)[:-1])
df['Postal code'] = df['Postal code'].map(lambda x: str(x)[:-1])
df['Neighborhood'] = df['Neighborhood'].map(lambda x: str(x)[:-1])
df.head(30)

Borough                                       Neighborhood  \
0       Not assigned                                                      
1       Not assigned                                                      
2         North York                                          Parkwoods   
3         North York                                   Victoria Village   
4   Downtown Toronto                         Regent Park / Harbourfront   
5         North York                  Lawrence Manor / Lawrence Heights   
6   Downtown Toronto       Queen's Park / Ontario Provincial Government   
7       Not assigned                                                      
8          Etobicoke                                   Islington Avenue   
9        Scarborough                                    Malvern / Rouge   
10      Not assigned                                                      
11        North York                                          Don Mills   
12         East York                   Parkview Hill / Woodbine Gardens   
13  Downtown Toronto                           Garden District, Ryerson   
14        North York                                          Glencairn   
15      Not assigned                                                      
16      Not assigned                                                      
17         Etobicoke  West Deane Park / Princess Gardens / Martin Gr...   
18       Scarborough           Rouge Hill / Port Union / Highland Creek   
19      Not assigned                                                      
20        North York                                          Don Mills   
21         East York                                   Woodbine Heights   
22  Downtown Toronto                                     St. James Town   
23              York                                 Humewood-Cedarvale   
24      Not assigned                                                      
25      Not assigned                                                      
26         Etobicoke  Eringate / Bloordale Gardens / Old Burnhamthor...   
27       Scarborough                Guildwood / Morningside / West Hill   
28      Not assigned                                                      
29      Not assigned                                                      

   Postal code  
0          M1A  
1          M2A  
2          M3A  
3          M4A  
4          M5A  
5          M6A  
6          M7A  
7          M8A  
8          M9A  
9          M1B  
10         M2B  
11         M3B  
12         M4B  
13         M5B  
14         M6B  
15         M7B  
16         M8B  
17         M9B  
18         M1C  
19         M2C  
20         M3C  
21         M4C  
22         M5C  
23         M6C  
24         M7C  
25         M8C  
26         M9C  
27         M1E  
28         M2E  
29         M3E

### Change column order

In [7]:
#change column order
df = df[["Postal code", "Borough", "Neighborhood"]] 
df.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                            
1         M2A      Not assigned                            
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

### Rename column

In [8]:
#rename column
df.rename(columns = {'Postal code':'PostalCode'}, inplace = True) 
df.head()


PostalCode           Borough                Neighborhood
0        M1A      Not assigned                            
1        M2A      Not assigned                            
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

### Filter rows with Borough "Not assigned"

In [9]:
#filter rows with Borough not assigned
print(df.shape)
df = df[df.Borough != "Not assigned"]
print(df.shape)
df = df.reset_index(drop=True)
print(df.shape)
df.head()

(180, 3)
(103, 3)
(103, 3)


PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

##### Notice the rows have been filtered as desired from 180 rows to 103 rows.

### Find Duplicates

In [10]:
dup1 = df[df.duplicated(['PostalCode'])]
dup1.shape
dup1.head()

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

#### These are no duplicate rows with same postal code.

#### As one postal code may contain more than one Neighborhood sepearted by "/".We need to repnace "/" with comma in Neighborhood column.

In [11]:
df['Neighborhood'] = df['Neighborhood'].str.replace('/',',')


#### Check if any neighbourhood which is not assigned.

In [12]:
df_x = df[df.Neighborhood == "Not assigned"]
df_x.head()

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

##### There is no such neighbourhood.

#### The Final output will like as follows as expected.

In [13]:
df.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                          Regent Park , Harbourfront  
3                   Lawrence Manor , Lawrence Heights  
4        Queen's Park , Ontario Provincial Government  
5                                    Islington Avenue  
6                                     Malvern , Rouge  
7                                           Don Mills  
8                    Parkview Hill , Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park , Princess Gardens , Martin Gr...  
12           Rouge Hill , Port Union , Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights

In [14]:
df.shape
x = df.shape[0]
x

103

In [15]:
print("Number of rows in dataframe : ", df.shape[0])

Number of rows in dataframe :  103


### We got our first dataframe as df which has PostalCode, borough and neighbourhood.

### Now, let's get our second dataframe.


#### We will get this through a csv file from the link.

In [18]:
df1 = pd.read_csv("http://cocl.us/Geospatial_data")
df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Check how many rows both dataframes have.

In [22]:
df.shape
df1.shape
print("Number of rows of df: " , df.shape[0])
print("Number of rows of df1: " , df1.shape[0])

Number of rows of df:  103
Number of rows of df1:  103


#### Notice both have equal rows.


### Now, rename the column of df1 to PostalCode.

In [24]:
df1.rename(columns = {'Postal Code':'PostalCode'}, inplace = True) 
df1.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

### Now, merge both dataframes df and df1 on column PostalCode.

In [25]:
df3 = pd.merge(df, df1, on='PostalCode')
df3.head()

PostalCode           Borough                                  Neighborhood  \
0        M3A        North York                                     Parkwoods   
1        M4A        North York                              Victoria Village   
2        M5A  Downtown Toronto                    Regent Park , Harbourfront   
3        M6A        North York             Lawrence Manor , Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

### Check how many rows in merged dataframe d3

In [27]:
df3.shape
print("Number of rows in merged dataframe d3 : " , df3.shape[0])

Number of rows in merged dataframe d3 :  103


### This is same as df and df1.

### This concludesthe second question in assignment.

### Thank You.